### PR TITLE
Simplify and stabilize softmax_cross_entropy test

### PR DESCRIPTION
**Pull request by okuta**
This PR reduces test time 30sec to 4sec.

### CODE REVIEW

**Comment by unnonouno** [comment]
jenkins test this please

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 1]
```diff
@@ -100,19 +107,16 @@ def check_forward(self, x_data, t_data, class_weight, use_cudnn='always'):
         testing.assert_allclose(
             loss_expect, loss_value, **self.check_forward_options)
 
-    @condition.retry(3)
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Why did you remove this line?

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 2]
```diff
@@ -370,19 +372,26 @@ def check_forward(self, x_data, t_data, class_weight):
             testing.assert_allclose(
                 loss_expect, li, **self.check_forward_options)
 
-    @condition.retry(3)
     def test_forward_cpu(self):
-        with chainer.using_config('use_cudnn', self.use_cudnn):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Why did you remove this argument?

Comment (okuta) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Because, this is cpu test.

Comment (unnonouno) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Is it better to check if `use_cudnn` does not affect its behavior event when cup is used?

**Comment by okuta** [comment]
CPU mode never use cuDNN.
I think that it clearly understands from the test function name and code.

**Comment by unnonouno** [comment]
No, we need to test if it "never" use cuDNN even if you believe it from the code.

**Comment by okuta** [comment]
I aceppted your comment.

The test got 9 times faster.
Items: 13641  -> 469
Time : 50.2s -> 5.7s

**Comment by Crissman** [comment]
@unnonouno Please approve to merge.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 3]
```diff
@@ -576,19 +569,18 @@ def test_double_backward_gpu(self):
...
-    'use_cudnn': ['always', 'auto', 'never'],
+    'weight_apply': [True],
 }) + testing.product({
-    'shape': [None, (2, 3), (2, 3, 2), (2, 3, 2, 2)],
+    'shape_ignore': [(None, None),  ((2, 3, 2), (0, 1))],
     'normalize': [True, False],
-    'ignore_index': [(0, 1)],
-    'dtype': [numpy.float16, numpy.float32, numpy.float64],
-    'weight_apply': [False, True],
-    'use_cudnn': ['always', 'auto', 'never'],
```
[the remark this answers is not in the document]
Comment (okuta) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: I fixed.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 4]
```diff
@@ -455,7 +451,6 @@ def test_double_backward_gpu(self):
 
 
 @testing.parameterize(*testing.product({
-    'use_cudnn': ['always', 'auto', 'never'],
```
[the remark this answers is not in the document]
Comment (okuta) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: I fixed.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 5]
```diff
@@ -576,19 +579,20 @@ def test_double_backward_gpu(self):
...
-    'weight_apply': [False, True],
-    'use_cudnn': ['always', 'auto', 'never'],
+    'weight_apply': [True],
+    'use_cudnn': ['always', 'never'],
 }) + testing.product({
-    'shape': [None, (2, 3), (2, 3, 2), (2, 3, 2, 2)],
+    'shape_ignore': [(None, None),  ((2, 3, 2), (0, 1))],
     'normalize': [True, False],
-    'ignore_index': [(0, 1)],
-    'dtype': [numpy.float16, numpy.float32, numpy.float64],
```
[the remark this answers is not in the document]
Comment (okuta) [reply tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: I think no.

**Comment by okuta** [comment]
I fixed.

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 6]
```diff
@@ -13,36 +13,38 @@
...
-    'weight_apply': [False, True],
-    'enable_double_backprop': [False, True],
     'label_dtype': [numpy.int32],
 }) + testing.product({
-    'shape': [None, (2, 3), (2, 3, 2), (2, 3, 2, 2)],
+    'shape_ignore': [(None, None), ((2, 3, 2, 2), (0, 1, 0))],
     'cache_score': [False],
-    'normalize': [True],
-    'ignore_index': [(0, 1)],
-    'dtype': [numpy.float16, numpy.float32, numpy.float64],
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: This fix removes test cases for float types except float32 and I feel it is a problem because a developer easily make a bug for other float dtypes when he/she writes a CUDA kernel which does not use generics.
Actually SCE is not one simple FunctionNode, but it dispatches process based on its argument. For example it makes completely different function node with `enable_double_backprop=True`  and `False` (see the source). So it has two function implementations. With and without `class_weight` it calls different kernels (it has two kernel codes in its implementation).
It is better to consider a policy to reduce the number of test cases.

**Comment by okuta** [comment]
I splited this PR to three PRs (this, #4554 and #4555).

This PR forces on `TestSoftmaxCrossEntropyDisableDoubleBackprop` and `TestSoftmaxCrossEntropyEnableDoubleBackprop`.

**Comment by okuta** [comment]
Thank you for merging related PRs.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by toslunar** [comment]
ping? @unnonouno

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by toslunar** [comment]
The last two Jenkins tests covered all the testcases.
